did_parser/src/parsed_did.rs

```diff
@@ -0,0 +1,84 @@
+use std::convert::TryFrom;
+use std::fmt::{Display, Formatter};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::DidUrl;
+use crate::{error::ParseError, utils::parse::parse_did_method_id, DidRange};
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Did {
+    did: String,
+    method: DidRange,
+    id: DidRange,
+}
+
+impl Did {
+    pub fn parse(did: String) -> Result<Self, ParseError> {
+        let (_, method, id) = parse_did_method_id(&did)?;
+
+        if id.end > did.len() {
+            return Err(ParseError::InvalidInput("Invalid DID"));
+        }
+
+        Ok(Self { did, method, id })
+    }
+
+    pub fn did(&self) -> &str {
+        self.did.as_ref()
+    }
+
+    pub fn method(&self) -> &str {
+        self.did[self.method.clone()].as_ref()
+    }
+
+    pub fn id(&self) -> &str {
+        self.did[self.id.clone()].as_ref()
+    }
+}
+
+impl TryFrom<String> for Did {
+    type Error = ParseError;
+
+    fn try_from(did: String) -> Result<Self, Self::Error> {
+        Self::parse(did)
+    }
+}
+
+impl Display for Did {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.did)
+    }
+}
+
+impl Serialize for Did {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.did())
+    }
+}
+
+impl<'de> Deserialize<'de> for Did {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let did = String::deserialize(deserializer)?;
+        Self::parse(did).map_err(serde::de::Error::custom)
+    }
+}
+
+impl TryFrom<&DidUrl> for Did {
+    type Error = ParseError;
+
+    fn try_from(did_url: &DidUrl) -> Result<Self, Self::Error> {
+        Self::parse(
+            did_url
+                .did()
+                .ok_or(Self::Error::InvalidInput("No DID provided in the DID URL"))?
+                .to_string(),
+        )
```

Comment:
Shouldn't there be a better way to convert? The DID is already parsed, so it's just a matter of whether there are path arguments and stuff like that. If the `DidUrl` doesn't contain any of the additional stuff that makes it a did url, then the parts could be converted to a `Did`, right?

Reply:
Are you suggesting to expose the fields of `DidUrl` for this?

Reply:
Through `pub(crate)`, yeah. Why not? Or simply a constructor method could be added that's crate internal, and the from impl could be near the `ParsedDidUrl` type so the fields can stay private.

Reply:
I think the overhead caused by so negligible it's not worth additional code or exposing the now private fields at all. But I will do as you wish.

Reply:
The crate public constructor is probably the best way to go about it, so the fields aren't touched.

I think it's highly worth it. Why not optimize this when it's easy to do so? The difference between parsing and not parsing the string is huge in the number of executed instructions.

Reply:
Again, I don't think this is worth either changing the visibility of the fields or adding additional code. Keeping a struct's trait implementations in the same module it's defined in also makes the code navigable, orderly and well-organized. 

And again, a call to `parse_did_method_id` has absolutely negligible performance cost.

But since you think it's highly worth it, I will change it according to your preference.